worker/src/task.rs

```diff
@@ -0,0 +1,600 @@
+                drop(tx);
+                return ();
+            }
+            let success = wrapped_run.await;
+            let out = if success.is_some() {
+                Ok(success.unwrap())
+            } else {
+                let cleanup_result = wrapped_cleanup.unwrap().await;
+                if cleanup_result.is_none() {
+                    panic!("Cleanup failed to run for task with id: {task_id}");
+                }
+                Err(cleanup_result.unwrap())
+            };
+            if tx.send(out).is_err() {
+                warn!("Task with task_id {task_id} output could not be sent: receiver dropped")
+            }
+            ()
+        });
+        let task = Task {
+            task_id,
+            tag: tag.clone(),
+            shutdown: shutdown_cloned.clone(),
+        };
+        {
+            let mut state = self.state.lock();
+            if !detached {
+                state.tasks.insert(task_id, task);
+            }
+        }
+        let state_cloned = self.state.clone();
+        TaskHandle {
+            id: task_id,
+            shutdown: shutdown_cloned,
+            join_handle: tokio::spawn(async move {
+                fut.await;
+                if !detached {
+                    let mut state = state_cloned.lock();
+                    state.tasks.remove(&task_id);
+                }
+                rx.try_recv()
+            }),
+            detached: false,
+        }
+    }
+    /// signals shutdown of this executor and any Clones
+    pub async fn shutdown_and_wait(&self) {
+        // hang up the channel which will cause the dedicated thread
+        // to quit
+        let mut shutdown_started = false;
+        {
+            let mut state = self.state.lock();
+            if state.task_tracker_state == TaskTrackerState::Running {
+                state.task_tracker_state = TaskTrackerState::ShutdownStarted;
+                shutdown_started = true;
+            }
+        }
+        if shutdown_started {
+            self.cancel_all_and_wait().await;
+            {
+                let mut state = self.state.lock();
+                state.task_tracker_state = TaskTrackerState::ShutdownComplete;
+            }
+        }
+    }
+    pub async fn stop_task(&self, task_id: u64) {
+        let task;
+        {
+            let mut state = self.state.lock();
+            task = state.tasks.get_mut(&task_id).cloned();
+        }
+        if task.is_none() {
+            error!("Task with task_id {task_id} doesn't exist!");
+        } else {
+            let shutdown = task.unwrap().shutdown;
+            shutdown.clone().shutdown();
+            shutdown.clone().wait_shutdown_complete().await;
+        }
+    }
+    pub async fn stop_tagged_tasks(&self, tag: String) {
+        let mut task_ids: Vec<u64> = vec![];
+        let mut futures: Vec<Pin<Box<dyn Future<Output = ()>>>> = vec![];
```

Comment:
I have a sense that the `<Box<dyn Future....` could be replaced by the [BoxFuture](https://docs.rs/futures/0.3.17/futures/future/type.BoxFuture.html)